### DISCUSSION TITLE
Why this error?

### DISCUSSION THREAD

**Discussion by frederikhors** [tokio-rs/axum#1316] [Q&A] [answered]
I'm creating a beautiful example for `axum-rust-embed-spa` and I think it can be published in examples directory if you agree ([we need an example like this](https://github.com/tokio-rs/axum/discussions/1309)).

It's not 100% yet. There are still a few things missing for great SPA hosting, but now I'm stuck in a refactoring process that produces an error I don't understand.

- The repo is: https://github.com/frederikhors/rust-embed-spa (the `main` branch works!)

- The commit with the refactoring is: https://github.com/frederikhors/rust-embed-spa/commit/fe0e120fee37b84205e4e8954450606b2efd35e7

- The error is:

```
mismatched types
expected opaque type `impl IntoResponse`
        found struct `Response<http_body::combinators::box_body::UnsyncBoxBody<axum::body::Bytes, axum::Error>>` rustc E0308
main.rs(49, 34): the expected opaque type
main.rs(32, 16): return type inferred to be `impl IntoResponse` here
```

![image](https://user-images.githubusercontent.com/41120635/186402388-7a19c256-0172-43e6-a6b0-aaa1603961b2.png)

Can you help me understand what's going on?

**Comment by davidpdrsn** [comment, marked as answer]
It boils down to this https://docs.rs/axum/latest/axum/response/index.html#returning-different-response-types

**Reply by frederikhors** [reply]
So I should simply write `.into_response()` here and there?

Like this?

```rust
async fn static_handler(uri: Uri) -> impl IntoResponse {
    let path = uri.path().trim_start_matches('/').to_string();

    if path.is_empty() || path == "index.html" {
        return get_index_response().await.into_response(); // <--------------------- THIS
    }

    match Assets::get(path.as_str()) {
        Some(content) => {
            let body = boxed(Full::from(content.data));
            let mime = mime_guess::from_path(path).first_or_octet_stream();

            Response::builder()
                .header(header::CONTENT_TYPE, mime.as_ref())
                .body(body)
                .unwrap()
        }
        None => get_index_response().await.into_response() // <--------------------- THIS
    }
}

async fn get_index_response() -> impl IntoResponse {
    match Assets::get("index.html") {
        Some(content) => {
            let body = boxed(Full::from(content.data));
            let mime = mime_guess::from_path("index.html").first_or_octet_stream();

            Response::builder()
                .header(header::CONTENT_TYPE, mime.as_ref())
                .body(body)
                .unwrap()
        }
        None => Response::builder()
            .status(StatusCode::NOT_FOUND)
            .body(boxed(Full::from("404")))
            .unwrap(),
    }
}
```

**Reply by frederikhors** [reply]
It works but I don't know if it's good.

**Reply by davidpdrsn** [reply]
> but I don't know if it's good

Its good

**Reply by davidpdrsn** [reply]
😅
